compiler/qsc_hir/src/ty.rs

```diff
@@ -61,28 +61,68 @@ impl Ty {
                     .map(|item| item.with_package(package))
                     .collect(),
             ),
-            Ty::Udt(res) => Ty::Udt(res.with_package(package)),
+            Ty::Udt(name, res) => Ty::Udt(name.clone(), res.with_package(package)),
+        }
+    }
+
+    pub fn display(&self) -> String {
+        match self {
+            Ty::Array(item) => {
+                format!("{}[]", item.display())
+            }
+            Ty::Arrow(arrow) => {
+                let arrow_symbol = match arrow.kind {
+                    CallableKind::Function => "->",
+                    CallableKind::Operation => "=>",
+                };
+                let functors = if arrow.functors == FunctorSet::Value(FunctorSetValue::Empty) {
+                    String::new()
+                } else {
+                    format!(" is {}", arrow.functors)
+                };
+                format!(
+                    "({} {arrow_symbol} {}{functors})",
+                    arrow.input.display(),
+                    arrow.output.display()
+                )
+            }
+            Ty::Infer(_) | Ty::Err => "?".to_string(),
+            Ty::Param(name, _) | Ty::Udt(name, _) => name.to_string(),
+            Ty::Prim(prim) => format!("{prim:?}"),
+            Ty::Tuple(items) => {
+                if items.is_empty() {
+                    "Unit".to_string()
+                } else if items.len() == 1 {
+                    let item = items.get(0).expect("expected single item");
+                    format!("({},)", item.display())
+                } else {
+                    let items = items.iter().map(Ty::display).collect::<Vec<_>>().join(", ");
+                    format!("({items})")
+                }
+            }
         }
     }
 }
 
 impl Display for Ty {
```

Comment:
I wonder if it's worth unifying this implementation with the `display` such that only `fmt` is needed, and then the places where `.display()` were called become `.to_string()` which internally calls `fmt`. Right now there are a few differences, but if we could reconcile them then there could be a single implementation rather than two that we need to remember to choose between.

Reply:
I considered trying to do this, but realized we really have separate needs for printing depending on the context. The HIR printing really needs to print all components of a struct in as compact a form as possible, while the error message/language server messages don't need to tell the user all the internal fields of the struct and needs to prioritize easy to read displays. Because they have such different needs, I'm not sure we can have one single way of printing these data structs that can satisfy both.

Reply:
As an aside though, I did briefly look into using formatting flags with the `fmt` impl to get the two different ways of printing to be used with `format!` macro, but that turned out to be a bit more cryptic than simply having a separate `.display` function.